File: app/src/main/java/com/acrcloud/ui/select/SelectMusicViewModel.java
package com.acrcloud.ui.select;

import android.databinding.ObservableArrayList;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;
import android.os.Environment;
import android.util.Log;

import com.acrcloud.ui.Song;
import com.acrcloud.ui.base.BaseViewModel;
import com.acrcloud.ui.edit.MainNavigator;
import com.acrcloud.utils.SongRenamer;

import java.io.File;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import hu.akarnokd.rxjava2.subjects.DispatchWorkSubject;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

public class SelectMusicViewModel extends BaseViewModel<MainNavigator> {

    public final DispatchWorkSubject<Song> openSongEvent = DispatchWorkSubject.create(getSchedulerProvider().ui());
    public final DispatchWorkSubject<Song> itemSongChangedEvent = DispatchWorkSubject.create(getSchedulerProvider().ui());

    public final ObservableField<String> folderPath = new ObservableField<>(Environment.getExternalStorageDirectory().toString());
    public final ObservableArrayList<Song> songs = new ObservableArrayList<>();
    private ObservableBoolean isEditing = new ObservableBoolean(false);
    private Disposable editingFolderDisposable;


    public SelectMusicViewModel() {

        openSongEvent.subscribe(song -> {
            if (getNavigator() != null) {
                getNavigator().onItemSongSelected(song);
            }
        });
    }

    public void loadFolder(String path) {


        if (path.equals("/")) {
            return;
        }

        this.folderPath.set(path);
        loadFolder();
    }

    public void loadFolder() {
        songs.clear();
        if (folderPath.get() != null) {
            File folder = new File(folderPath.get());
            if (folder.exists() && folder.listFiles() != null) {
                if (folder.listFiles().length != 0) {
                    for (File file : folder.listFiles()) {

                        if (file.isDirectory()) {
                            Song song = new Song(file.getAbsolutePath(), file.getName());
                            song.setType(Song.TYPE.FOLDER);
                            songs.add(song);
                        } else if (isMusicFile(file.getAbsolutePath())) {
                            songs.add(new Song(file.getAbsolutePath(), file.getName()));
                        }
                    }
                }
            }
        }
    }

    public void toParentFolder() {
        File curFolder = new File(folderPath.get());

        //TODO costylik
        if (curFolder.getParent().endsWith("/emulated")) {
            loadFolder(curFolder.getParentFile().getParent());
            return;
        }

        loadFolder(curFolder.getParent());
    }

    //mp3, mp4, wav, m4a, aac, amr, ape, flv, flac, ogg, wma, caf, alac
    private boolean isMusicFile(String absolutePath) {
        return absolutePath.endsWith(".mp3") || absolutePath.endsWith(".mp4") || absolutePath.endsWith(".wav") || absolutePath.endsWith(".m4a") || absolutePath.endsWith(".aac") || absolutePath.endsWith(".amr") || absolutePath.endsWith(".ape") || absolutePath.endsWith(".flv") || absolutePath.endsWith(".flac") || absolutePath.endsWith(".ogg") || absolutePath.endsWith(".wma") || absolutePath.endsWith(".caf") || absolutePath.endsWith(".alac");
    }


    public void stopEditCurrentFolderNow() {
        if (isEditing.get()) {
            if (editingFolderDisposable != null) {
                editingFolderDisposable.dispose();
            }

            isEditing.set(false);
        }
    }

    public void editCurrentFolderNow() {

        if (isEditing.get()) {
            return;
        }

        isEditing.set(true);

        ExecutorService poolExecutor = Executors.newFixedThreadPool(1);
        Scheduler scheduler = Schedulers.from(poolExecutor);
        editingFolderDisposable = Observable.fromIterable(songs)
                .filter(song -> song.getType() == Song.TYPE.SONG)
                .flatMap((Function<Song, ObservableSource<Song>>) song ->
                        getDataManager().recognizeSong(song.getPath())
                                .subscribeOn(scheduler)
                                .doOnSubscribe((d) -> song.setIsEditing(true))
                                .doOnNext(r -> song.setIsEditing(false))
                                .map(result -> {
                                    Song edited = SongRenamer.editSong(song, result);

                                    if (edited != null) {

                                        updateSongInLists(edited, song);

                                        if (songs.indexOf(song) == songs.size() - 1) {
                                            isEditing.set(false);
                                        }

                                        return edited;
                                    }

                                    return song;

                                }))
                .doOnDispose(new Action() {
                    @Override
                    public void run() throws Exception {
                        for (Song item : songs) {
                            item.setIsEditing(false);
                        }
                    }
                })
                .subscribe(newSong -> {
                            getMessage().onNext(newSong.getTitle());
                        },
                        throwable -> {
                            Log.e("error", "renameAll: ", throwable);
                        });
    }

    public void onItemSelected(Song song) {
        if (song.getType() == Song.TYPE.SONG) {
            openSongEvent.onNext(song);
        } else {

            String path = song.getPath();

            //TODO costylik
            if (path.endsWith("/emulated")) {
                path = path.concat("/0");
            }

            loadFolder(path);
        }
    }

    public void editSelectedSongNow(Song song) {

        getDataManager().recognizeSong(song.getPath())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe((d) -> song.setIsEditing(true))
                .doOnNext(r -> song.setIsEditing(false))
                .subscribe(result -> {
                    Song edited = SongRenamer.editSong(song, result);

                    if (edited != null) {
                        updateSongInLists(edited, song);
                    }

                }, throwable -> {
                    Log.e("error", "rename: ", throwable);
                });

    }

    private void updateSongInLists(Song newSong, Song oldSong) {
        int index = -1;

        for (Song s : songs) {
            if (s.getPath().equals(oldSong.getPath())) {
                index = songs.indexOf(s);
                break;
            }
        }
        if (index > -1) {
            songs.set(index, newSong);
        }
    }

    public void editSelectedFolderNow(Song song) {

    }

    public ObservableBoolean getIsEditing() {
        return isEditing;
    }
}

File: app/src/main/java/com/acrcloud/ui/select/SelectMusicViewModel.kt
package com.acrcloud.ui.select

import android.databinding.ObservableArrayList
import android.databinding.ObservableBoolean
import android.databinding.ObservableField
import android.os.Environment
import android.util.Log
import com.acrcloud.ui.EditSong
import com.acrcloud.ui.base.BaseViewModel
import com.acrcloud.ui.edit.MainNavigator
import com.acrcloud.utils.SongEditor
import hu.akarnokd.rxjava2.subjects.DispatchWorkSubject
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.io.File
import java.util.concurrent.Executors

class SelectMusicViewModel : BaseViewModel<MainNavigator>() {

    val openSongEvent = DispatchWorkSubject.create<EditSong>(schedulerProvider.ui())
    val itemSongChangedEvent = DispatchWorkSubject.create<EditSong>(schedulerProvider.ui())

    val folderPath = ObservableField(Environment.getExternalStorageDirectory().path)
    val files = ObservableArrayList<EditSong>()
    val isEditingFolder = ObservableBoolean(false)
    private var editingFolderDisposable: Disposable? = null

    init {

        openSongEvent.subscribe { song ->
            if (navigator != null) {
                navigator!!.onItemSongSelected(song)
            }
        }

        loadItems()
    }

    private fun loadItems() {
        files.clear()
        if (folderPath.get() != null) {
            val folder = File(folderPath.get())
            if (folder.exists() && folder.listFiles() != null) {
                if (folder.listFiles().size != 0) {
                    for (file in folder.listFiles()) {

                        if (file.isDirectory) {
                            val song = EditSong(file.absolutePath, file.name)
                            song.type = EditSong.TYPE.FOLDER
                            files.add(song)
                        } else if (isMusicFile(file.absolutePath)) {
                            files.add(EditSong(file.absolutePath, file.name))
                        }
                    }
                    files.sortBy { it.type }
                }
            }
        }
    }

    fun toParentFolder() {
        val curFolder = File(folderPath.get())
        if (curFolder.parentFile != null && curFolder.parentFile.canRead()) {
            folderPath.set(curFolder.parent)
            checkAndLoadItems()
        }
    }

    private fun checkAndLoadItems() {
        val curFolder = File(folderPath.get())

        if (curFolder != null || curFolder.canRead()) {
            loadItems()
        }
    }

    //mp3, mp4, wav, m4a, aac, amr, ape, flv, flac, ogg, wma, caf, alac
    private fun isMusicFile(absolutePath: String): Boolean {
        return absolutePath.endsWith(".mp3") || absolutePath.endsWith(".mp4") || absolutePath.endsWith(".wav") || absolutePath.endsWith(".m4a") || absolutePath.endsWith(".aac") || absolutePath.endsWith(".amr") || absolutePath.endsWith(".ape") || absolutePath.endsWith(".flv") || absolutePath.endsWith(".flac") || absolutePath.endsWith(".ogg") || absolutePath.endsWith(".wma") || absolutePath.endsWith(".caf") || absolutePath.endsWith(".alac")
    }


    fun stopEditCurrentFolderNow() {
        if (isEditingFolder.get()) {
            if (editingFolderDisposable != null) {
                editingFolderDisposable!!.dispose()
            }

            isEditingFolder.set(false)
        }
    }

    fun editCurrentFolderNow() {

        if (isEditingFolder.get() && files.any { it.type == EditSong.TYPE.SONG }) {
            return
        }

        isEditingFolder.set(true)

        val poolExecutorSubscribe = Executors.newFixedThreadPool(1)
        val schedulerSubscribe = Schedulers.from(poolExecutorSubscribe)

        editingFolderDisposable = Observable.fromIterable(files)
                .filter { song -> song.type == EditSong.TYPE.SONG }
                .flatMap { editSong: EditSong ->
                    dataManager.recognizeSong(editSong.path)
                            .subscribeOn(schedulerSubscribe)
                            .doOnSubscribe { d -> editSong.isEditing = true }
                            .doOnNext { r -> editSong.isEditing = false }
                            .map<EditSong> { result ->
                                val edited = SongEditor.editSong(editSong, result)

                                if (edited != null) {

                                    updateSongInLists(edited, editSong)

                                    //last song
                                    if (files.indexOf(editSong) == files.size - 1) {
                                        isEditingFolder.set(false)
                                    }

                                    edited
                                }

                                editSong

                            }
                }
                .doOnDispose {
                    for (item in files) {
                        item.isEditing = false
                    }
                }
                .subscribe({ newSong -> message.onNext(newSong.title!!) },
                        { throwable -> Log.e("error", "renameAll: ", throwable) })
    }

    fun onItemSelected(editSong: EditSong) {
        if (editSong.type == EditSong.TYPE.SONG) {
            openSongEvent.onNext(editSong)
        } else {

            var folder = editSong.path
            if (folder != null) {
                folderPath.set(folder)
                checkAndLoadItems()
            }
        }
    }

    fun editSelectedSongNow(editSong: EditSong) {

        dataManager.recognizeSong(editSong.path)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe { d -> editSong.isEditing = true }
                .doOnNext { r -> editSong.isEditing = false }
                .subscribe({ result ->
                    val edited = SongEditor.editSong(editSong, result)

                    if (edited != null) {
                        updateSongInLists(edited, editSong)
                    }

                }, { throwable -> Log.e("error", "rename: ", throwable) })

    }

    private fun updateSongInLists(newEditSong: EditSong, oldEditSong: EditSong) {
        var index = -1

        for (s in files) {
            if (s.path == oldEditSong.path) {
                index = files.indexOf(s)
                break
            }
        }
        if (index > -1) {
            files[index] = newEditSong
        }
    }

    fun editSelectedFolderNow(editSong: EditSong) {

    }
}
